Add support for Pluggable Discoveries (#1333)
arduino/arduino-cli: docs/platform-specification.md

- `{build.variant.path}`: The path to the selected board variant folder (inside the
  [variant platform](#platform-terminology), for example hardware/arduino/avr/variants/micro)

If the platform supports pluggable discovery it may also declare a set of `upload_port.*` properties, these properties
will be used to identify a board by the discovery process when plugged in.

> For example we could declare a series of `upload_port.vid` and `upload_port.pid` for the Uno like so:

Reviewer: ```suggestion
For example we could declare a series of `upload_port.vid` and `upload_port.pid` properties for the Uno like so:
```

    uno.upload_port.vid.0=0x2341
    uno.upload_port.pid.0=0x0043
    uno.upload_port.vid.1=0x2341
    uno.upload_port.pid.1=0x0001
    uno.upload_port.vid.2=0x2A03
    uno.upload_port.pid.2=0x0043
    uno.upload_port.vid.3=0x2341
    uno.upload_port.pid.3=0x0243

> In this case we're using the board's `vid` and `pid` to identify it but `upload_port.*` properties can be anything that
> can help identify a certain board. For more detailed information see the

Reviewer: ```suggestion
In this case we're using the board's USB VID/PID pair to identify it but `upload_port.*` properties can be anything that
can help identify a certain board. For more detailed information see the
```
The platform properties are named with `vid` and `pid`, but the "vid" and "pid" in this sentence are referring to the USB vendor ID (VID) and product ID (PID), which the reader may not be familiar with. This change will at least provide them with some additional search terms. I wanted to linkify the text when I was working on https://arduino.github.io/arduino-cli/dev/platform-specification/#board-vidpid, but I didn't find a straightforward authoritative explanation to link to.

variables.

#### Pluggable Discovery

Discovery tools are a special kind of tools that are used to find supported boards, a platform must declare one or more
Pluggable Discoveries in its `platform.txt`. It's ok to use another platform discovery or the builtin ones if necessary.

There are two different syntaxes to declare discoveries, if the platform uses just one discovery:

```
discovery.required=PLATFORM:DISCOVERY_NAME
```

instead if it needs multiple discoveries:

```
discovery.required.0=PLATFORM:DISCOVERY_ID_1
discovery.required.1=PLATFORM:DISCOVERY_ID_2
```

A platform that supports only boards connected to the serial can easily use the builtin `serial-discovery` without
creating a custom Pluggable Discovery:

```
discovery.required=builtin:serial-discovery
```

if it supports also boards connected to the network it can use the builtin `mdns-discovery`:

```
discovery.required.0=builtin:serial-discovery
discovery.required.1=builtin:mdns-discovery
```

Since the above syntax requires adding a discovery in the `discoveryDependencies` field in the platform
`package_index.json` it might be cumbersome to do it before releasing it so we also provide another syntax to ease

> development:

Author: It should be possible to keep both with different IDs but they would both be started by the CLI if they're in the `platform.txt`, either one must be omitted or commented out.

> development:

Reviewer: OK. I was thinking that if there was a way for them to coexist then we might want to find a way to explain that, since I foresee this will be something of interest to platform developers. Since there isn't, I think the system is explained well enough.

property can be used as **{cmd.path}** inside the recipe, and the same happens for all the other avrdude configuration
variables.

> #### Pluggable Discovery

Reviewer: ```suggestion
#### Pluggable discovery
```
Even though it could be argued that [title case](https://en.wikipedia.org/wiki/Title_case) rules should apply here, most of the document's section headings use sentence case and I think that it's best to follow the established style convention.

Discovery tools are a special kind of tool used to find supported boards. A platform must declare one or more Pluggable
Discoveries in its [`platform.txt`](#platformtxt). Discoveries can be referenced from other packages, including the
`builtin` dummy package which contains the traditional discoveries.

There are two different syntaxes to declare discoveries. If the platform uses just one discovery:

    pluggable_discovery.required=VENDOR_ID:DISCOVERY_NAME

instead if it needs multiple discoveries:

    pluggable_discovery.required.0=VENDOR_ID:DISCOVERY_0_NAME
    pluggable_discovery.required.1=VENDOR_ID:DISCOVERY_1_NAME

A platform that supports only boards connected via serial ports can easily use the `builtin` package's

> `serial-discovery` without creating a custom Pluggable Discovery:

Reviewer: ```suggestion
`serial-discovery` without creating a custom pluggable discovery:
```

[`arduino-cli upload`](commands/arduino-cli_upload.md). Arduino uses the term "upload" for the process of transferring a
program to the Arduino board.

The **upload.tool.<protocol_name\>** property determines the tool to be used for upload. A specific
**upload.tool.<protocol_name\>** property should be defined for every board in boards.txt:

    [......]
    uno.upload.tool.serial=avrdude
    [......]
    leonardo.upload.tool.serial=avrdude
    leonardo.upload.tool.network=arduino_ota
    [......]

When the user tries to upload using a certain protocol but the board doesn't support it it will fallback to `default` if
specified. A board can always specify a `default` protocol even if others are defined:

    [......]
    uno.upload.tool.default=avrdude
    [......]
    leonardo.upload.tool.default=avrdude
    leonardo.upload.tool.network=arduino_ota
    [......]

`default` is also used when no upload address is provided by the user, this can be done only for boards that use upload
tools with builtin port detection like `openocd`.

> For backward compatibility with IDE 1.8.15 and older the previous syntax is still supported:

Author: We can't know exactly which version of all products will contain this feature though, I think my also be misleading if we reference versions that are still not released. Probably we should update this when new versions are released.

> For backward compatibility with IDE 1.8.15 and older the previous syntax is still supported:

Reviewer: Yeah, it's really not a great situation. My concern is that the need to add the information will be forgotten. But guessing the versions risks the introduction of incorrect information.

I think that this will be very important information for platform developers using pluggable discovery because they will need to clearly document the minimum supported tool versions.

    tools.bossac.upload.network_pattern="{runtime.tools.arduinoOTA.path}/bin/arduinoOTA"
                                        -address {serial.port} -port 65280
                                        -sketch "{build.path}/{build.project_name}.bin"

with:

    tools.arduino_ota.upload.pattern="{runtime.tools.arduinoOTA.path}/bin/arduinoOTA"
                                    -address {upload.port.address} -port 65280
                                    -sketch "{build.path}/{build.project_name}.bin"

#### User provided fields

Some upload recipes might require custom fields that must be provided by the user, like username and password to upload
over the network. In this case the recipe must use the special placeholder {upload.field.FIELD_NAME} where FIELD_NAME
must be declared separately in the recipe using the following format:

    tools.UPLOAD_RECIPE_ID.upload.field.FIELD_NAME=FIELD_LABEL
    tools.UPLOAD_RECIPE_ID.upload.field.FIELD_NAME.secret=true

FIELD_LABEL is the label shown in the graphical prompt to ask the user to enter the value of the field. The secret
property is optional and it should be set to true if the field is a secret (like passwords or token).

> Let’s see how a complete example will look like:

Reviewer: ```suggestion
Let's see a complete example:
```

    tools.arduino_ota.upload.field.username=Username
    tools.arduino_ota.upload.field.password=Password
    tools.arduino_ota.upload.field.password.secret=true

>     tools.arduino_ota.upload.pattern="{runtime.tools.arduinoOTA.path}/bin/arduinoOTA"
>                                     -address {upload.port.address} -port 65280
>                                     -username "{upload.field.username}
>                                     -password "{upload.field.password}"
>                                     -sketch "{build.path}/{build.project_name}.bin"

Reviewer: ```suggestion
    tools.arduino_ota.upload.pattern="{runtime.tools.arduinoOTA.path}/bin/arduinoOTA" -address {upload.port.address} -port 65280 -username "{upload.field.username} -password "{upload.field.password}" -sketch "{build.path}/{build.project_name}.bin"
```
It was reasonable in the special case where you wanted to make it easy to see the before and after and made an explicit note about it being broken for clarity, but I think it will cause a lot of confusion to platform authors to do that in other parts of the specification, since the properties data format doesn't provide any way to break things into multiple lines AFAIK.

Most **{upload.XXXX}** variables are used later in the avrdude upload recipe in platform.txt:

    [.....]
    tools.avrdude.upload.pattern="{cmd.path}" "-C{config.path}" {upload.verbose} -p{build.mcu} -c{upload.port.protocol} -P{upload.port.address} -b{upload.speed} -D "-Uflash:w:{build.path}/{build.project_name}.hex:i"
    [.....]

If necessary the same property can be defined multiple times for different protocols:

    leonardo.upload.serial.maximum_size=28672
    leonardo.upload.network.maximum_size=256

The two above properties will be available as **{upload.serial.maximum_size}** and **{upload.network.maximum_size}**.

> #### Properties from Pluggable Discovery

Reviewer: ```suggestion
#### Properties from pluggable discovery
```

> If a platform supports Pluggable Discovery it can also use the port's properties returned by a discovery. For example,

Reviewer: ```suggestion
If a platform supports pluggable discovery it can also use the port's properties returned by a discovery. For example,
```

compiled sketch to a board using an external programmer.

The **program.tool** property determines the tool to be used for this action. This property is typically defined for

> each programmer in [programmers.txt](#programmerstxt) and uses the same syntax as `upload` action:

Reviewer: ```suggestion
each programmer in [programmers.txt](#programmerstxt) and uses the same syntax as
[the `upload` action](#sketch-upload-configuration):
```

1. `bootloader` is used to flash the bootloader to the board

The **bootloader.tool** property determines the tool to be used for the `erase` and `bootloader` actions both. This

> property is typically defined for each board in boards.txt and uses the same syntax as `upload` action:

Reviewer: ```suggestion
property is typically defined for each board in boards.txt and uses the same syntax as
[the `upload` action](#sketch-upload-configuration):
```

    [......]
    uno.bootloader.tool.serial=avrdude
    [......]

>     leonardo.upload.tool.serial=avrdude
>     leonardo.upload.tool.network=arduino_ota
>     [......]
>     duemilanove.upload.tool.default=avrdude

Reviewer: ```suggestion
    leonardo.bootloader.tool.serial=avrdude
    leonardo.bootloader.tool.network=arduino_ota
    [......]
    duemilanove.bootloader.tool.default=avrdude
```

  This behavior
  [can be configured](https://arduino.github.io/arduino-cli/latest/commands/arduino-cli_core_install/#options)
- **Arduino Pro IDE**: (since 0.1.0) runs the script for any installed platform.

> 
> ```
> 
> ```

Reviewer: ```suggestion
```
